Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2Connections.swift

```diff
@@ -0,0 +1,532 @@
+            let availableStreams = self.connections[index].connected(connection, maxStreams: maxConcurrentStreams)
+            let context = AvailableConnectionContext(
+                availableStreams: availableStreams,
+                eventLoop: connection.eventLoop,
+                isIdle: self.connections[index].isIdle
+            )
+            return (index, context)
+        }
+
+        /// Move the HTTP1ConnectionState to backingOff.
+        ///
+        /// - Parameter connectionID: The connectionID of the failed connection attempt
+        /// - Returns: The eventLoop on which to schedule the backoff timer
+        mutating func backoffNextConnectionAttempt(_ connectionID: Connection.ID) -> EventLoop {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connectionID }) else {
+                preconditionFailure("We tried to create a new connection that we know nothing about?")
+            }
+
+            self.connections[index].failedToConnect()
+            return self.connections[index].eventLoop
+        }
+
+        // MARK: Connection lifecycle events
+
+        mutating func goAwayReceived(_ connectionID: Connection.ID) {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connectionID }) else {
+                preconditionFailure("go away recieved for a connection that does not exists")
+            }
+            self.connections[index].goAwayReceived()
+        }
+
+        mutating func newHTTP2MaxConcurrentStreamsReceived(
+            _ connectionID: Connection.ID,
+            newMaxStreams: Int
+        ) -> (Int, AvailableConnectionContext) {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connectionID }) else {
+                preconditionFailure("We tried to update the maximum concurren streams number for a connection that does not exists")
+            }
+            let availableStreams = self.connections[index].newMaxConcurrentStreams(newMaxStreams)
+            let context = AvailableConnectionContext(
+                availableStreams: availableStreams,
+                eventLoop: self.connections[index].eventLoop,
+                isIdle: self.connections[index].isIdle
+            )
+            return (index, context)
+        }
+
+        // MARK: Leasing and releasing
+
+        mutating func leaseStream(onPreferred eventLoop: EventLoop) -> Connection? {
+            guard let index = self.findAvailableConnection(onPreferred: eventLoop, minimumAvailableStreams: 1) else { return nil }
+            return self.leaseStreams(at: index, count: 1)
+        }
+
+        /// tries to find an available connection on the prefered `eventLoop`. If it can't find one with the given `eventLoop`, it returns the first available connection
+        private func findAvailableConnection(onPreferred eventLoop: EventLoop, minimumAvailableStreams: Int) -> Int? {
```

Comment:
When would `minimumAvailableStreams` not be `1`? This function would in my mind only be called, when a new request was just created. In this case we always only require one stream.

Reply:
Rename `findAvailableStream`

Reply:
But we are actually searching for a connection and return a connection and not a stream.